source/fab/engine.py

```diff
@@ -0,0 +1,190 @@
+                        # Are the dependencies compiled?
+                        if discovery[dependency] == "Compiled":
+                            compiled[idep] = True
+                    else:
+                        # If the dependency isn't in the list at all yet
+                        # then add an entry so the system knows we are
+                        # expecting it later (for the above check)
+                        new_discovery[dependency] = "HeardOf"
+
+                # If the dependencies are satisfied (or there weren't
+                # any) then the file can be compiled now
+                if len(compiled) == 0 or all(compiled):
+                    for definition in artifact.defines:
+                        task = self._taskmap[(artifact.filetype,
+                                              artifact.state)]
+                        new_artifacts.extend(task.run([artifact]))
+                        new_discovery[definition] = "Compiled"
+                else:
+                    # If the dependencies weren't all satisfied then
+                    # back on the queue for another pass later
+                    new_artifacts.append(artifact)
+            else:
+                # If it wasn't required it could be later, so
+                # put it back on the queue, unless the target
+                # has been compiled, in which case it wasn't
+                # needed at all!
+                if discovery[self._target] != "Compiled":
+                    new_artifacts.append(artifact)
+
+        elif artifact.state is Compiled:
+            # Begin populating the list for linking
+            new_objects.append(artifact)
+            # But do not return a new artifact - this object
+            # is "done" as far as the processing is concerned
+
+            # But, if this is the file containing the target
+            # that means everything must have been compiled
+            # by this point; so we can do the linking step
+            if self._target in artifact.defines:
+                task = self._taskmap[(artifact.filetype,
+                                      artifact.state)]
+                new_artifacts.extend(task.run(objects + [artifact]))
+
+        elif artifact.state is Linked:
+            # Nothing to do at present with the final linked
+            # executable, but included here for completeness
+            pass
+
+        else:
+            # An artifact with a filetype and state set
+            # will have an appropriate task that should
+            # be used to run it (though unlike the old
+            # implementation this is probably returning
+            # the instance of the Task not the class)
+            if ((artifact.filetype, artifact.state)
+                    in self._taskmap):
+                task = self._taskmap[(artifact.filetype,
+                                      artifact.state)]
+
+                new_artifacts.extend(task.run([artifact]))
```

Comment:
Rather than the big compound `if` tree it may be possible to streamline this using a dictionary. An optimisation for subsequent consideration.

Reply:
It's definitely ripe for splitting up a bit - if `pylint` were here I expect it would be very upset.  I was thinking _maybe_ rather than a dictionary it may be possible to just use specific names - the branches which are more involved are _mostly_ describing actions which I hope are fairly language/configuration agnostic (e.g. one always has to do some sort of walk of the analysed objects and track compiles).  But yes either way let's not worry about it right now!